File: leetcode/054-spiral-matrix/SpiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new LinkedList<>();
        
        int m = matrix.length;
        
        if (m == 0) {
            return result;
        }
        
        int n = matrix[0].length;
        
        int borderLeft = 0;
        int borderRight = n - 1;
        int borderBottom = m - 1;
        int borderTop = 0;

        while (borderLeft <= borderRight && borderTop <= borderBottom) {
            for (int l = borderLeft; l <= borderRight; l++) {
                result.add(matrix[borderTop][l]);
            }
            
            borderTop++;
            
            for (int l = borderTop; l <= borderBottom; l++) {
                result.add(matrix[l][borderRight]);
            }

            borderRight--;
            
            
            if (borderTop <= borderBottom) {
                for (int l = borderRight; l >= borderLeft; l--) {
                    result.add(matrix[borderBottom][l]);
                }
            }
            
            borderBottom--;
            
            if (borderLeft <= borderRight) {
                for (int l = borderBottom; l >= borderTop; l--) {
                    result.add(matrix[l][borderLeft]);
                }
            }
            
            borderLeft++;
        }
        
        return result;
    }
}

File: leetcode/054-spiral-matrix/spiral-matrix.js
const spiralOrder = matrix => {
    const result = [];

    const r = matrix.length;

    if (r === 0) {
        return result;
    }

    const c = matrix[0].length;

    let n = r * c;

    let minX = 0, minY = 0, maxX = c - 1, maxY = r - 1;

    while (n > 0) {
        for (let i = minX; i <= maxX && n > 0; i++) {
            result.push(matrix[minY][i]);
            n--;
        }

        minY++;

        for (let i = minY; i <= maxY && n > 0; i++) {
            result.push(matrix[i][maxX]);
            n--;
        }

        maxX--;

        for (let i = maxX; i >= minX && n > 0; i--) {
            result.push(matrix[maxY][i]);
            n--;
        }

        maxY--;

        for (let i = maxY; i >= minY && n > 0; i--) {
            result.push(matrix[i][minX]);
            n--;
        }

        minX++;
    }

    return result;
};
